[9.0][2nd revolution] port to v9 with several improvements (sorrento code sprint)

alexis-via opened the pull request:
I started to port bank-payment to v9 (with a lot of improvements) during the Sorrento Code sprint 2016.

Improvements include:
- full re-organisation of modules and big re-organisation of the code
- simplification of the code related to the fact that support for direct debit is now in t
  he base module, not added by an optional module
- new design of the wizard to select move lines to pay
- support for non-SEPA file transfer-
- support for German direct debit SEPA files (fixes #129)
- remove workflow

This port to v9 is not finished... there is still a lot of work:
- finish the code of account_payment_order/wizard/account_payment_line_create.py
- port account_banking_payment_transfer and integrate it inside account_payment_order
- fix bugs
- clean-up code, remove dead code
- test in several complex scenarios

rlizana committed:
[IMP] Add filter group by payment mode in account invoices

rlizana committed:
[FIX] Remove deprecated tag type

alexis-via committed:
Add bank.payment.lines object to allow grouping in the payments

alexis-via committed:
Allow to select customer refunds in direct debit order and supplier refunds in payment order (to reduce the amount to pay)

alexis-via committed:
Remove storno, which seems specific to one particular country.

alexis-via committed:
Add a hook to inherit grouping of the transfer account move line

Use that new hook in SEPA direct debits
Better variable names

alexis-via committed:
Remove dead XML code

alexis-via committed:
FIX Reading wrong field for sequence type of SEPA DD

Loop on bank payment lines instead of payment lines

pedrobaeza committed:
[FIX] account_banking_payment_export: Correct communication type for customer invoices in payment lines

hbrunn committed:
[ADD] bump version, migrate total column

pedrobaeza committed:
[FIX] account_banking_payment_export: Avoid the drop of the column total on updates

alexis-via committed:
Forgot to remove a usecase of debit_denied

Add ACL

pedrobaeza committed:
Merge pull request #219 from pedrobaeza/8.0-account_banking_payment_export-migrate_payment_order_total

[8.0] account_banking_payment_export: migrate payment.order total

pedrobaeza committed:
[FIX] account_banking_payment_export: Check if column exists before creating it

sbidoul committed:
[UPD] addons table in README.md

alexis-via committed:
Add constraints on BIC length (8 or 11 chars)

alexis-via committed:
Post transfer account move by default, like for an invoice

Add sum in bank payment line tree view

alexis-via committed:
Better checks and nice error messages

alexis-via committed:
FIX better handling of account_id in transfer move

alexis-via committed:
Add bank_line_count field and display it in view

Display date_sent field in view

pedrobaeza committed:
[FIX] account_payment_purchase: Propagate payment_mode on MTO/drop shipping

sbidoul committed:
Merge pull request #227 from pedrobaeza/8.0-account_payment_purchase-mto_fix

[FIX] account_payment_purchase: Propagate payment_mode on MTO/drop shipping

alexis-via committed:
Better filters on payment.order.create wizard

Add default values for those filters on payment.mode

alexis-via committed:
The module account_banking is not used anymore ! Hourra !

alexis-via committed:
FIX remove coma at end of line

alexis-via committed:
Update automated tests and demo data

Add on_change on field 'type' of payment.mode for easier configuration

alexis-via committed:
Update test in account_banking_tests

Commit:
[IMP][account_payment_blocking] Allow to set no follow up on draft invoices

sbidoul committed:
[FIX] account_payment_mode_term new api compatibility

Without this patch we have `TypeError: extend_payment_order_domain() takes exactly 3 arguments (2 given)` when adding an invoice to a payment order.

pedrobaeza committed:
Merge pull request #234 from OCA/8.0-patch-account_payment_mode_term-sbi

[FIX] account_payment_mode_term new api compatibility

OCA-git-bot committed:
[UPD] addons table in README.md

Commit:
account_payment_blocking : migrate payment order create to new api

pedrobaeza committed:
[FIX] account_payment_*: Ensure work with dropshipping dual invoice

When invoicing from dropshipping, the same picking serves for both
invoicing sale and purchase orders. The OCA module
stock_dropshipping_dual_invoice makes that this happens, but
as the picking is the same, it contains both the sale and the
purchase reference, making that the resulting invoices don't have
the correct payment mode. This is harmless for the rest of
operations, as the dictionary query is provided with the proper
default value. On v9, dropshipping dual invoice has been merged
into core, so this patch for sure will be welcome in that version
(if the used method is the same).

pedrobaeza committed:
Merge pull request #238 from pedrobaeza/8.0-account_payment-fix_dropshipping

[FIX] account_payment_*: Ensure work with dropshipping dual invoice

alexis-via committed:
Add translatable 'label' field on payment.mode, designed to be used in invoice report

alexis-via committed:
Better strings

pedrobaeza committed:
[FIX] account_banking_mandate: Fix name_get call

antespi committed:
[FIX] Allow to select bank account when creating a mandate

pedrobaeza committed:
Merge pull request #240 from Antiun/fix-mandate_creation

[8.0][FIX][account_banking_mandate] Allow to select bank account when creating a mandate

sergio-teruel committed:
[NEW][8.0] portal_payment_mode: Adds ACL's to payment_mode

pedrobaeza committed:
[FIX] account_banking_mandate: es translation

DarioLodeiros committed:
Mandates multicompany error #242 

The restriction rule Banking Mandate multi-company, gives an error when editing a partner bank account from a company A when that account has created a mandate from another company B. So it is impossible to create a mandate for each company without disabling the rule.
Solution: Uncheck the rule writte in Banking Mandate multi-company. But i'm not sure this is correct.

pedrobaeza committed:
Merge pull request #241 from incaser/8.0-portal_payment_mode

[NEW][8.0] portal_payment_mode. Adds ACL's to payment_mode model for portal users

pedrobaeza committed:
Merge pull request #248 from StefanRijnhart/8.0-remove_account_banking

The module account_banking is not used anymore ! Hourra !

pedrobaeza committed:
[IMP] *: Short headers

pedrobaeza committed:
[IMP] account_payment_partner: Remove label field

There's already a field in the payment mode called 'note' that is
printed on the invoices, so there's no need of another field for
that purpose. This field is added by account_banking_payment_export

pedrobaeza committed:
[IMP] *: Bump version numbers

pedrobaeza committed:
[IMP] account_banking_payment_transfer: Restore entry_posted check

pedrobaeza committed:
[IMP] *: es translations

OCA-git-bot committed:
[UPD] addons table in README.md

pedrobaeza committed:
[FIX] bank_statement_instant_voucher: PEP8

pedrobaeza committed:
[IMP] account_banking_payment_export: Make journal_ids not required

Now if it's empty, all journals are taken into account

pedrobaeza committed:
[IMP] account_banking_payment_export: default for "invoice" field in wizard

pedrobaeza committed:
[FIX] account_banking_payment_export: Fix move lines domain in create payment order

pedrobaeza committed:
[FIX] account_direct_debit: Fix move lines domain for direct debits

pedrobaeza committed:
[IMP] account_banking_payment_export: Improve computed method + store field

StefanRijnhart committed:
Merge pull request #250 from pedrobaeza/8.0-bank-payment-lines

[8.0] banking: Add support for grouping the payments, for deducing a refund from an invoice and remove storno

OCA-git-bot committed:
[UPD] addons table in README.md

sbidoul committed:
camel case

sbidoul committed:
fix previous commit

pedrobaeza committed:
[IMP] account_banking_payment_export: Add option for not grouping lines

There can be multiple reasons for not grouping payment lines:

* The customer needs each payment line individually for matching references.
* For avoiding the risk of returning a big amount.
* For not overpassing amout limits.
* ...

So with this commit, you have the possibility of selecting at payment mode
level if you want to group or not

pedrobaeza committed:
Merge pull request #254 from pedrobaeza/8.0-payment_order-grouping_options

[IMP] account_banking_payment_export: Add option for not grouping lines

OCA-git-bot committed:
[UPD] addons table in README.md

pedrobaeza committed:
Merge pull request #243 from DarioLodeiros/patch-1

 Mandates multicompany error #242

pedrobaeza committed:
Merge pull request #217 from pedrobaeza/8.0-reference_type-fix

[FIX] account_banking_payment_export: Correct communication type for customer invoices in payment lines

pedrobaeza committed:
Merge pull request #71 from rlizana/8.0_payment_mode_filter

[IMP] Add filter group by payment mode to invoices

pedrobaeza committed:
[IMP] account_payment_partner: Name to filter for better inheratiblity

pedrobaeza committed:
Merge pull request #237 from acsone/8.0-fix_payment_order_create-lga

account_payment_blocking : migrate payment order create to new api

Commit:
[CHG][account_payment_blocking] Bump version

pedrobaeza committed:
Merge pull request #232 from acsone/8.0-imp-payment-blocking-draft-invoice-ape

[IMP][account_payment_blocking] Allow to set no follow up on draft invoices

OCA-git-bot committed:
[UPD] addons table in README.md

antespi committed:
[IMP] Define SEPA identifiers per payment mode

pedrobaeza committed:
[IMP] .travis.yml: Coveralls activation

pedrobaeza committed:
Merge pull request #256 from Antiun/imp-sepa_per_payment_mode

[8.0][IMP][account_banking_*] Define SEPA identifiers per payment mode

OCA-git-bot committed:
[UPD] addons table in README.md

pedrobaeza committed:
[FIX] account_banking_sepa_direct_debit: Fixes #257

sergio-teruel committed:
[FIX][8.0] account_banking_mandate: Add creditor identifier field to report mandate.

[FIX][8.0] account_banking_mandate: Translate.

sergio-teruel committed:
[FIX][8.0] account_banking_mandate: Slipt basic mandate and sepa.

sergio-teruel committed:
[IMP][8.0] sepa_direct_debit_mandate: Split basic and sepa mandate

sergio-teruel committed:
[IMP][8.0] account_banking_sepa_direct_debit: Use api.multi instead

sergio-teruel committed:
[IMP][8.0] account_banking_mandate: Use api.multi instead

pedrobaeza committed:
[IMP] .travis.yml: Include Transifex

sergio-teruel committed:
[IMP][8.0] sepa_direct_debit_mandate: Migration script

sergio-teruel committed:
[IMP][8.0] account_banking_mandate: Make format field required

Commit:
[FIX][account_payment_blocking] In some cases, there isn't only one line on receivable/payable account on journal entry genereted from invoice.

sergio-teruel committed:
[IMP][8.0] account_banking_mandate: Pep8 fixes

sergio-teruel committed:
[IMP][8.0] account_banking_sepa_direct_debit: Pep8 fixes

OCA-git-bot committed:
[ADD] setup.py

sergio-teruel committed:
[FIX][8.0] account_banking_sepa_direct_debit: Fix translation

pedrobaeza committed:
Merge pull request #264 from acsone/8.0-fix-payment-blocking-term-ape

[FIX][account_payment_blocking] In some cases, there isn't only one ine on receivable/payable account on journal entry genereted from invoice.

sergio-teruel committed:
[IMP][8.0] account_banking_sepa_direct_debit: Report font size smallest

pedrobaeza committed:
Merge pull request #265 from sergio-teruel/8.0-PR-Adaptar-mandate

[IMP][8.0] account_banking_mandate:

OCA-git-bot committed:
[UPD] addons table in README.md

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

schout-it committed:
[8.0][BUG][account_banking_payment_transfer]

_create_move_line_partner_account raise Usererror
2 variables, only 1 used

pedrobaeza committed:
Merge pull request #267 from schout-it/8.0

[8.0][BUG][account_banking_payment_transfer]

oca-transbot committed:
OCA Transbot updated translations from Transifex

alexis-via committed:
Merge remote-tracking branch 'origin/8.0' into 9-sorrento-code-sprint

alexis-via committed:
Start to port bank-payment to v9 (with a lot of improvements) during the Sorrento Code sprint 2016

Improvements include:
- full re-organisation of modules and big re-organisation of the code
- simplification of the code related to the fact that support for direct debit is now in t
he base module, not added by an optional module account_direct_debit (module was removed)
- new design of the wizard to select move lines to pay
- support for non-SEPA file transfer-
- support for German direct debit SEPA files (fixes bug #129)
- remove workflow of payment.order

This port to v9 is not finished... there is still a lot of work:
- finish the code of account_payment_order/wizard/account_payment_line_create.py
- port account_banking_payment_transfer and integrate it inside account_payment_order
- fix bugs
- clean-up code, remove dead code
- test in several complex scenarios

alexis-via committed:
Port account_payment_sale to v9

yvaucher commented on account_banking_mandate/models/account_invoice.py:
```diff
@@ -1,33 +1,27 @@
+        if line.get('type') == 'dest' and line.get('invoice_id'):
+            invoice = self.browse(line['invoice_id'])
+            if invoice.type in ('out_invoice', 'out_refund'):
+                res['mandate_id'] = invoice.mandate_id.id or False
```
`or False` is unecessary

yvaucher commented on account_banking_mandate/models/account_move_line.py:
```diff
@@ -0,0 +1,23 @@
+        # TODO : test on the view field "mandate required ?"
+        if payment_order.payment_type == 'inbound' and self.mandate_id:
+            vals['mandate_id'] = self.mandate_id.id or False
+            vals['partner_bank_id'] = self.mandate_id.partner_bank_id.id or False
```
`or False` is unecessary

yvaucher commented on account_banking_mandate/models/account_payment_line.py:
```diff
@@ -0,0 +1,63 @@
+            if partner_bank_id and 'mandate_id' not in vals:
+                mandates = self.env['account.banking.mandate'].search(
+                    [('partner_bank_id', '=', partner_bank_id),
+                     ('state', '=', 'valid')])
```
You can add `limit=1` as you get the first mandate in any case.

yvaucher commented on account_banking_mandate/models/account_payment_line.py:
```diff
@@ -0,0 +1,63 @@
+                    vals['mandate_id'] = mandates[0].id
+        return super(AccountPaymentLine, self).create(vals)
+
+    @api.one
```
`api.one` is deprecated

yvaucher commented on account_banking_mandate/views/account_move_line.xml:
```diff
@@ -0,0 +1,26 @@
+-->
+
+<openerp>
+<data>
```
You can use `<odoo>`

yvaucher commented on account_banking_pain_base/models/account_payment_mode.py:
```diff
@@ -0,0 +1,50 @@
+        "- Country code (2, optional)\n"
+        "- Company idenfier (N, VAT)\n"
+        "- Service suffix (N, issued by bank)")
+    # I plan to change this -- Alexis
```
Still valid? at least add ? TODO

Can it be removed ?

alexis-via replied:
In some countries (spain I think), for some strange reasons, if the company has several banks, they may have different initiating_party_issuer (one for each bank). Some Spanish guys could confirm this.

yvaucher commented on account_payment_order/models/account_payment_line.py:
```diff
@@ -0,0 +1,124 @@
+    @api.onchange('partner_id')
+    def partner_id_change(self):
+        partner_bank = False
+        if self.partner_id and self.partner_id.bank_ids:
```
You can remove `self.partner_id and`.

If `self.partner_id` is `False`
then `self.partner_id.bank_ids` will return `False`

yvaucher commented on account_payment_order/wizard/account_payment_line_create.py:
```diff
@@ -77,12 +91,41 @@ def extend_payment_order_domain(self, payment_order, domain):
+    def populate(self):
+        domain = self._prepare_move_line_domain()
+        lines = self.env['account.move.line'].search(domain)
+        self.move_line_ids = [(6, 0, lines.ids)]
```
No need to pass a list with magic number of relation, you can simply assign `self.move_line_ids = lines`

alexis-via commented:
@yvaucher Thanks for your review and your remarks. This PR is still a work in progress, so the code will continue to change, in particular in the selection of move lines wizard and with the merge of account_banking_payment_transfer in account_payment_order.

yvaucher commented on account_payment_order/wizard/account_payment_line_create.py:
```diff
@@ -224,23 +217,11 @@ def _prepare_payment_line(self, payment, line):
+                self.order_id)
+        return True
         # Force reload of payment order view as a workaround for lp:1155525
         return {'name': _('Payment Orders'),
```
This code is unreachable

yvaucher commented:
@alexis-via Couldn't be possible to split it in multiple PR ? It's quite hard to review.

Would be great to have a first PR with only the `account_payment_order` module.

yvaucher commented:
@alexis-via I added the WIP tag on it.

yvaucher commented:
The README.rst of `account_payment_order` needs some rewriting.

It still says it depends on account_payment

alexis-via committed:
Finalise the wizard of selection of move lines to pay

Add button "Add to payment/debit order" on invoice form view
Started to integrate payment transfer in account_payment_order (not finished at all though)
Various fixes/changes/improvements/...

alexis-via committed:
Remove module account_payment_sale_stock

alexis-via commented:
@pedrobaeza  @sbidoul  With the new way of creating supplier invoices in v9 (manual creation of the supplier invoice then select purchase order to add to invoice), I think that the module account_payment_purchase should be dropped in v9. Do you agree ?

sbidoul commented:
@alexis-via or you could copy the payment info when adding the first PO, and warn if the next one differs... it becomes complex to understand for the user, though.

Ok to drop it AFAIC.

alexis-via committed:
bank_bic on res.partner.bank is a related field in v9 and is not displayed in form view, so no need for explainations any more

alexis-via committed:
Update and re-enable demo data

alexis-via committed:
Finalise merge of account_banking_payment_transfer into account_payment_order

Add support for transfer moves
Display transfer moves in a dedicated tab in payment order view
Code cleanup

alexis-via committed:
Remove module account_payment_purchase

alexis-via commented:
@sbidoul @pedrobaeza 
Now that we dropped the workflow on the payment order object, it is more difficult to make the payment order automatically move to "done" state when all the transfer moves are reconciled with the lines in the bank journal.

I propose another approach in v9 for this: remove the 'done' state ; the last state would be "File Uploaded". Add a function field on payment.order that tell you if the transfer moves are fully reconciled or not. For that, we could imagine a "fa-" type button that would display for example "2/5 transfer moves reconciled", and, if you click on this button, it would show the related transfer move lines.

alexis-via committed:
Take into account the review of @yvaucher

alexis-via committed:
PEP8 compliant with the latest version of flake8

alexis-via committed:
Fix tests of account_banking_mandate

alexis-via committed:
Filter on allowed_journal_ids even if onchange of payment_mode_id is not played (for example when the payment line is pushed from the invoice)

alexis-via committed:
Add option 'target_move' (all or posted) in wizard to select move lines to pay

Remove module account_payment_include_draft_move

alexis-via committed:
Copy payment mode and mandate from invoice to refund

Run dos2unix on the few files that had DOS carriage return

alexis-via committed:
Add module account_banking_mandate_sale

Add option 'mandate_required' on payment orders
Autoselect first valid mandate on customer invoice that have a payment mode 'mandate_required' = True
Add option on select move lines to pay wizard to allow selection of litigation moves (unchecked by default), in order to integrate the feature of the module account_payment_blocking

alexis-via committed:
Add _id suffix on M2O fields customer_payment_mode and supplier_payment_mode (Odoo did the same on all the M2O property fields of res.partner)

alexis-via commented:
@adrienpeiffer FYI, I integrated the feature of your module account_payment_include_draft_move inside the module account_payment_order, so I deleted your module account_payment_include_draft_move in v9.

I also added an option "allow_blocked" in the wizard of the selection of move lines to pay... so maybe we could also delete your module "account_payment_blocking"... but this module has some additional invoice-related features. Do you agree if I delete the module account_payment_blocking from OCA/bank-payment/9.0 and you re-create a module account_invoice_blocking in OCA/account-invoicing/9.0 ?

pedrobaeza commented:
@alexis-via, I didn't use the state 'done' in my workflows. User didn't understand correctly when exactly it's put in it (remember that we use a lot of intermediate accounts to transfer the balance until an invoice is considered to be paid), so it's good for me to drop that state.

alexis-via committed:
Remove 'done' state on payment orders

sbidoul commented:
@alexis-via I liked the done state. But it is also easy to look at the unreconciled moves on the transfer account.

BTW, have you tried using the bank account as transfer account and use the new matching feature in bank statement reconciliation?

alexis-via commented:
@sbidoul
About the old "done" state, we could add a button or a function field to easily show the reconciled/unreconciled move on the transfer account... I think it would be a good remplacement.

 I haven't tried the use the "bank account" as transfer accout, because I don't know how the new "matching feature" for bank statement reconciliation works. Could you tell me a few words about it (or point me to some online source of info about it) and I'll test my PR with it.

sbidoul commented:
@alexis-via I've not looked at the code for the matching feature, but you can test it easily on runbot.
- open a vendor bill
- click register payment and select the bank journal -> a move is generated (debit supplier, credit bank), and the invoice is reconciled
- create a bank statement for the same amount and partner
- reconcile the bank statement: you see a blue line proposed which the line already existing on the bank account

So no need for a transfer account in such a scenario. I assume it will work for batch payments too.

alexis-via commented:
OK, thanks for your explaination. It's enough information for me to make some tests :)

alexis-via committed:
Payment date should never be in the past.

alexis-via committed:
Add support for non-SEPA DD and CT without without IBANs

alexis-via committed:
Move fields mandate_required and export_ascii from payment mode to payment method

Display chatter on payment orders
Several small usability improvements

alexis-via committed:
Avoid empty group in payment mode

Code cleanup in views

alexis-via committed:
Add number of mandates in partner form (similar to number of bank accounts)

alexis-via commented:
I discovered today that, in v9, you can link a bank account to a contact and not only to a parent partner. In v8, bank accounts could only be linked to a parent partner. I think the behavior of v8 was much better on this point, what do you think ? Shouldn't we restrict the right to link to a bank account only to parent partners ?

pedrobaeza commented:
I don't see the application of allowing to link a bank account to a contact, so I agree to restrict it.

yvaucher commented:
@alexis-via extracting account_payment_order module from this PR in another PR would unlock some other portage such as:
https://github.com/OCA/l10n-switzerland/pull/212

alexis-via commented:
I still have some important changes to make in the datamodel of the module account_payment_order, to support what Stéphane suggested in a previous comment: use the bank account as transfer account (it requires a change in the datamodel of the payment mode, because we should not ask for the transfer journal/account in this case).

After this change (I should do it in a few days), we can merge the whole PR.

alexis-via committed:
Add support for using bank account as offsetting account in journal entry

alexis-via committed:
FIX use of communication type

alexis-via committed:
Add option no_debit_before_maturity on payment mode

alexis-via committed:
Add ICS in account settings page

alexis-via committed:
No creation of bank accounts linked to contacts

alexis-via commented:
@sbidoul It now works with the transfer move that uses the bank account (and not a special transfer account).

alexis-via commented:
@pedrobaeza @sbidoul I blocked the creation/display of bank accounts for contacts. If you look at the domain on that line of code https://github.com/odoo/odoo/blob/9.0/openerp/addons/base/res/res_bank.py#L61 it is clear that it is a bug in Odoo and that you should not have a res.partner.bank linked to a contact.

elicoidal commented on account_banking_mandate/__openerp__.py:
```diff
@@ -1,29 +1,13 @@
     'name': 'Account Banking Mandate',
     'summary': 'Banking mandates',
-    'version': '8.0.0.1.0',
+    'version': '9.0.0.2.0',
```
9.0.1.0.0

elicoidal commented on account_banking_mandate/__openerp__.py:
```diff
@@ -32,21 +16,22 @@
-        'reports/account_banking_mandate_view.xml',
-        'reports/account_banking_mandate.xml',
     ],
     'demo': [],
```
Please remove empty keys

elicoidal commented on account_banking_mandate/models/account_banking_mandate.py:
```diff
@@ -1,154 +1,134 @@
+    _sql_constraints = [(
+        'mandate_ref_company_uniq',
+        'unique(unique_mandate_reference, company_id)',
+        'A Mandate with the same reference already exists for this company !')]
```
remove the space before "!" (multiple)

elicoidal commented on account_banking_mandate/models/account_banking_mandate.py:
```diff
@@ -1,154 +1,134 @@
+        for mandate in self:
+            if mandate.state != 'draft':
+                raise UserError(
+                    _('Mandate should be in draft state'))
```
period (several)

elicoidal commented on account_banking_mandate/models/account_payment_method.py:
```diff
@@ -0,0 +1,14 @@
+
+    mandate_required = fields.Boolean(
+        string='Mandate Required',
+        help="Activate this option is this payment method requires your "
```
s/is/if

elicoidal commented on account_banking_mandate/models/res_partner_bank.py:
```diff
@@ -1,34 +1,16 @@
+    mandate_ids = fields.One2many(
+        comodel_name='account.banking.mandate', inverse_name='partner_bank_id',
+        string='Direct Debit Mandates',
+        help='Banking mandates represents an authorization that the bank '
```
s/represents/represent

elicoidal commented on account_banking_mandate_sale/README.rst:
```diff
@@ -0,0 +1,69 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Account Banking Mandate Sale
```
Add a line  before (same length)

elicoidal commented on account_banking_mandate_sale/README.rst:
```diff
@@ -0,0 +1,69 @@
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/bank-payment/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed `feedback
```
update this section with newest template content

elicoidal commented on account_banking_pain_base/README.rst:
```diff
@@ -30,38 +31,42 @@ Usage
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/bank-payment/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed `feedback
```
use newest template content

elicoidal commented on account_banking_pain_base/models/account_payment_line.py:
```diff
@@ -0,0 +1,23 @@
+        ('NORM', 'Normal'),
+        ('HIGH', 'High')],
+        string='Priority', default='NORM',
+        help="This field will be used as the 'Instruction Priority' in "
```
remove 'the'

elicoidal commented on account_banking_pain_base/models/account_payment_method.py:
```diff
@@ -0,0 +1,24 @@
+    pain_version = fields.Selection([], string='PAIN Version')
+    convert_to_ascii = fields.Boolean(
+        string='Convert to ASCII', default=True,
+        help="If active, Odoo will convert each accented caracter to "
```
s/caracter/character (several)

elicoidal commented on account_banking_pain_base/views/account_payment_method.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data>
+
```
you could add copyright in all xml files

guewen replied:
AFAIK we did never put copyrights in XML files. I think we should try to not be too nitpicky, for it can drag down motivation and contributions.

guewen replied:
I read "you should" and not "you could",  your message is indeed different.

elicoidal commented on account_banking_sepa_credit_transfer/README.rst:
```diff
@@ -1,6 +1,6 @@
     :alt: License: AGPL-3
-    
+
 Account Banking SEPA Credit Transfer
```
add line above

elicoidal commented on account_banking_sepa_credit_transfer/README.rst:
```diff
@@ -32,45 +32,55 @@ This module is part of the OCA/bank-payment suite.
 =============
 
-No specific configuration.
+Create a Payment Mode dedicated to SEPA Credit Transfer and select
```
please use enumerated lists 
"#.create..." 
"#.Select"
etc.

elicoidal commented on account_banking_sepa_credit_transfer/data/account_payment_method.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data noupdate="1">
+
```
is there any reason for 2 empty lines before and after?

alexis-via committed:
Take into accounts the remarks of Eric Caudal

guewen commented:
I removed the work in progress flag.

alexis-via commented:
To make things clear : for me, this PR is ready to merge.

alexis-via commented:
@guewen By the way, can I have the rights to add/remove flags ? I can't do that for the moment...

alexis-via committed:
name of payment mode is now translatable

pedrobaeza commented:
@alexis-via, it's supposed that PSC members can do it. Aren't you member of the PSC?

guewen commented:
> @guewen By the way, can I have the rights to add/remove flags ? I can't do that for the moment...

Members of the PSC or those part of the OCA maintainers team only can add or remove flags. This is in my opinion one of the worst limitation of the github pull requests. Still, do not mind to ask someone to change a flag (the ideal would be to have a bot to whom we could ask those actions).

sbidoul commented:
I just added @alexis-via in banking maintainers. He is already part of the project on the OCA Odoo instance, I thought it was synchornized automatically.

guewen commented:
> To make things clear : for me, this PR is ready to merge.

:+1: for an optimistic merge on this PR

sbidoul commented:
`<parenthesis>`
First let me share a thought about this optimistic merging meme: don't forget it goes with a whole process (http://rfc.zeromq.org/spec:22), including test coverage, release management, evolution of public contract/api etc. So that's not something that must be used out of context without caution. 

In particular, OCA users expect that when something is merged on the main branch, it is stable. Merging means releasing to the world. I think that's something that must change at some point, but that's the situation we have today.
`</parenthesis>`

When this PR is merged, can people trust the data model will be stable or migration scripts will be written if it needs to evolve? I'm not worried for myself yet as I don't need this before a few month, but that's the usual expectation of OCA users. That said, I think the current data model in this PR is very good. @alexis-via do you think it's stable and can be installed in production? Otherwise, it should be merged with a big warning in the module README's.

Then (Alexis's going to hate me for writing this), I think the test coverage is a bit on the low side. I've not looked in detail but maybe it's not too hard to add some more tests that cover a bit more of the logic (not all the UI code)?

guewen commented:
@sbidoul I want to avoid misinterpretation about my mention of the optimistic merging, I was talking about this very pull request only. I concur with your words about the optimistic merging. In the current situation, I would not be in favor of using it on the OCA's projects for many reasons (amongst which the lack of test coverage, not a practical usage of versions numbers, probably not much people who would take the time to correct the red builds which is already such a battle with pessimistic merging currently, ...). `<troll>`_By the way, the optimistic merging advocate Pieter Hintjens also says that a feature without tests should be removed, it would be a large cleaning of the OCA's addons._`</troll>`

Now, I think that sometimes the rules are too rigid and that the judgement of one or a few enlightened reviewers (disclaimer: I'm certainly not one for the Accounting projects) should be able to overstep the rules in individual cases.

For this PR, I'm not entitled to judge but here is my feeling: these addons are the core part for many other usages (localizations, automation of sales (`sale_automatic_workflow`) and e-commerce at least) and are awaited. Once merged, people will be able (or at least more easily able) to work on their own modules extending the ones of this PR and then will surely fix or improve them along the way. On the contrary, if we leave this large and intimidating PR as-is, my fear is to have a stalling situation.

Another intermediate solution would have been to split the PR as I started with #274 but it was not pleasing @alexis-via.

> That said, I think the current data model in this PR is very good. @alexis-via do you think it's stable and can be installed in production? Otherwise, it should be merged with a big warning in the module README's.

Right, if the datamodel is not stable yet, we should probably not merge it yet.
A warning would be great in any case.

Well, the decision here is yours (the Accounting PSC) :)

alexis-via commented:
I confirm that, for me, the datamodel is stable (I would not say that this PR is ready to merge if I felt that the datamodel needed some changes) and I like it very much. This datamodel has been extensively discussed during the sorrento code sprint with Stephane and Pedro, and it only received some small updates/changes since the code sprint.

I made a several tests with both direct debits and credit transfer, and it worked well. My tests were not on a production database with some real data, but I tried to make it as real as possible. I think it is neither more nor less buggy than the rest of the Odoo accounting code base :) But at least we have clean and easy code, which is a big step forward.

I haven't ported the tests to v9. It's a big work because they need to be re-written, just like this port to v9 which is almost a rewrite, at least for the lower-level modules. Also, I think it's a pain to write tests on accounting in v9 because several objects (journals, accounts) that had XMLIDs in v8 don't have XMLIDs in v9 any more. For example, the port of the donation modules to v9 is unfinished because I didn't succeed to port the tests for that reason, cf the discussion on https://github.com/OCA/donation/pull/13

sbidoul commented:
@alexis-via it's indeed a pain to not have xmlids on journals anymore. In mis_builder I resolved to recreate everything from scratch for my tests (company, journal, accounts, etc). It turned out to be very easy and has the advantage of making the test independent of the rest of demo data. Example here: https://github.com/OCA/account-financial-reporting/blob/851a01570f821d054857e2c5f50e25b02c1d5145/mis_builder/tests/test_aep.py#L18

pedrobaeza commented:
For my tests development, I'm always recreating lately all the needed records to not depend on the variability of the demo data. It's not too hard.

sbidoul commented:
@guewen 

> I want to avoid misinterpretation about my mention of the optimistic merging

No misinterpretation from my side, don't worry. I just took the opportunity to try and land a point.

> On the contrary, if we leave this large and intimidating PR as-is, my fear is to have a stalling situation.

Agreed.

So let's move on. Ideally, it should go in a kind of beta stage, but we don't have that concept now and it's not in this PR that we'll improve OCA processes.

Still... tests, tests, tests. When we have good test coverage, it's so much easier to review further changes. To a large extent it's also the lack of tests that slow things down.

cyrilgdn commented:
> For my tests development, I'm always recreating lately all the needed records to not depend on the variability of the demo data. It's not too hard.

Agreed! Furthermore, for having migrated several modules in V9, it's much less painful to understand, fix or extend existing tests when all necessary data are created in tests setup (or in a common TestCase for sharing these data between several test cases) .

Commit:
[ADD] setup.py

adrienpeiffer commented:
I propose to add setup.py in https://github.com/akretion/banking/pull/1

Commit:
Merge pull request #1 from acsone/9.0-bp-setup-ape

[ADD] setup.py

alexis-via commented:
Good news: I got hired to port "account_banking_tests" to v9 ; I plan to do it for June 7th.

pedrobaeza commented:
Cool! Please include the test in each of the modules instead of having an specific module.

alexis-via commented:
@pedrobaeza I was planning to keep Stefan's approach with a module dedicated to tests, that would be above all the modules and allow to test to whole thing in the same code, which makes it easier to develop and more readable.

pedrobaeza commented:
But it's also worst to detect 2 things:
- The source of the problem
- If one module installed lonely gives any error.

Obviously I'm not thinking only on Travis, but on customer deployments with discrete installed modules and possible customized dependent modules. A test module is a nightmare for these things. So please reconsider again to make them separately with the corresponding tests on each one. Maybe some module like pain_base doesn't make sense, but now that we have a big generic module, it's even easier to make the tests inside the module.

fclementic2c commented:
I could eventually do some tests on this PR and I could generate a SEPA xml file :)

But I have several issues that I would like to discuss: 
- I cannot pay the balance of an invoice paid partially from 'Create payment lines from move lines' in payment order (it works from 'Add to payment order' button on a partially reconciled invoice)
- I cannot see multi-currency amounts when I add account move lines manually to payment order. The default view is useless (the pop-up window is too narrow so I cannot see amount currency, I see move lines with bank accounts - should only be payable accounts lines visible and I do not want to see a debit or credit columns but the balance left to be paid) 
- I cannot re-open a payment order after file as been uploaded (see my use case no1). The whole workflow should be 'cancel-able' I think but it is feasible ?

use case no1
- I get this message '_Cannot generate the account move when the currency of the payment (GBP) is not the same as the currency of the company (USD). This is not supported for the moment._' when I click on file successfully uploaded and I pay an invoice in GBP wheras my journal is in USD AND my receivable account is forced in GBP (I think this is the last point which generate this wrong message - probably an issue to be reported to odoo). However, I am stuck here since I cannot cancel my payment order since the file has been generated.

**Improvement suggestions**
- Most of users will try to pay supplier invoice from the menu supplier invoice and I have the feeling that we are very close to handle this. So I suggest 2 small improvements that will improve greatly the user experience I think.
  1 - Add the payment mode in the vendor invoices tree view (between the 'To pay' and 'Status' column)
  2 - Add the action 'Add to payment order' in tree view also 
- In the payment order, the button 'Create payment lines from move lines' should be renamed for a better end-user understanding; for ex. 'Select Items to pay' ; 
- A tool tip should explain what is the purpose of the bank payment lines (ex : lines grouped by date, partner...) 
- Bank and bank account should be mandatory in journal according to payment methods (it took me 40 minutes to find out why I couldn't generate an SEPA xml file whereas my bank account in my company was already set.)
- In the same spirit, on the invoice, the bank account should be mandatory according to payment methods.
- Invoiced are not paid at the end of the payment order workflow whereas payment are in the ledger. It might confusing since in the bank st. reconcile view you can match the payable line or the bank payment line in blue... At some point, it will be good to use the advance reconcile module maybe (already ported to v9) so you will be able to launch a reconcile on all payable account move lines related to the payment order.

*_Bug (?) found *_
- If you create a bank account from the company and then try to add it to a journal then a message tells you that can cannot re-create a bank account with the same number.
- If you select a contact of a vendor. the invoice will not add the bank account of the partner parent.
- View Bank Payment line : error KeyError: u'partner_id' when trying to group partner

Voila
Thanks

Frederic

alexis-via committed:
Add hooks for l10n_ch_sepa (pain.001.001.03.ch.02)

Move code of account_banking_pain_base/models/banking_export_pain.py to account_banking_pain_base/models/account_payment_order.py
Allow cancel when payment order is in "File Generated" state
Add postal address in party block

fclementic2c commented:
Functional tests Ok (improvement suggested above can be done in a separate PR)
:+1:

alexis-via committed:
Adds flexibility to support direct debit direct debit for switzerland (pain.008.001.02.ch.01)

Implement SEPA Core recommandations for postal address

alexis-via commented:
FYI, in order to properly implement SEPA for switzerland, I plan to add a field "local instrument" on payment line and bank payment line in account_banking_pain_base (this is a generic field, because local instrument is part of the ISO 20022 standard). I will add this field in the coming days. I hope it will be the last change in the datamodel for this PR.

sbidoul committed:
remote obsolete setup.py files

sbidoul committed:
pep8 long line

Commit:
Merge pull request #2 from acsone/setup-sbi

remove obsolete setup.py files that came from the 8.0 branch

alexis-via committed:
Add field local_instrument in payment lines and bank payment lines

alexis-via committed:
Merge branch '9-sorrento-code-sprint' of github.com:akretion/banking into 9-sorrento-code-sprint

alexis-via commented:
I made the change I wanted to make in the data model, with the addition of local_instrument on the payment line. I hope it is really the last change in the datamodel now ! It should be confirm with the tests that will be done on l10n_ch_sepa very soon.

alexis-via committed:
Don't group payment lines that use structured communication

alexis-via committed:
Add local instrucment in PmtInfId to warranty unicity

alexis-via committed:
Add the bank_line as argument to some methods that generate SEPA XML

alexis-via committed:
Move the code that generate the bank account number in the SEPA XML file in a dedicated method

alexis-via committed:
Simpler variable names

Rename method

alexis-via committed:
Add wizard to select several invoices at once and add them to payment order

Fix bug in groupby on bank payment lines

alexis-via committed:
Store field acc_type on res.partner.bank, so that we can search and groupby on it

alexis-via committed:
Better handling of partner_bank_id and mandate_id on invoice: invisible/required/onchange

alexis-via committed:
Fix crash in mandate report

sbidoul committed:
Avoid confusion between payment mode and payment method

sbidoul commented:
@alexis-via is it possible that currently the manual payment methods prevent finishing the payment order flow?

alexis-via committed:
Move field bank_account_required from module account_payment_partner to account_payment_mode

Make the mandate a required field on payment line when the payment method has mandate_required=True
Make the bank account a required field on payment line when the payment method has bank_account_required=True
Minor code cleanup
PEP8

alexis-via commented:
@sbidoul I don't understand your last question. Do you mean that you found an issue that you want  me to fix, or do you want to block the manual payment methods ?

sbidoul commented:
@alexis-via I could not finish the payment order flow for manual_out method. I'm not sure if it's me missing something, or a missing feature, or something else.

BTW, I finally found some time to test, and I must say what you have achieved here is exceptional! I love the configuration possibilities of the new payment modes.

alexis-via committed:
FIX manual payment method

alexis-via committed:
Enable the payment methods by default on bank journals (including existing bank journals via post_install scripts)

Enable "Generate Accounting Entries On File Upload" by default

alexis-via commented:
@sbidoul I fixed the issue with the manual payment method in my last commit.

I am very glad that you like my work on the 2nd revolution in bank-payment. You participated in the architecture of this new datamodel during the sorrento code sprint, so you can also congratulate yourself :)

alexis-via committed:
Add a unit test

alexis-via committed:
Better test that include several invoices (to test grouping) and refunds

alexis-via committed:
Add unittest for SEPA direct debit

alexis-via commented:
I added some unit tests in the modules account_banking_sepa_credit_transfer and account_banking_sepa_direct_debit. The test in account_banking_sepa_credit_transfer uses 2 suppliers with several invoices (to test grouping) and a refund (to test the deduction of a refund). Of course, these tests are still "modest" compared to the functional coverage of the bank-payment modules... I let you decide if they are enough to merge this PR or not.

fclementic2c commented:
@alexis-via 
I am still worry about Payment journal entries. Supplier invoices are not paid at the end of the workflow. In version 8 it was ok but for version 9 it is not adapted anymore. Can't we use the v9 Odoo payment object ?
(Saying that I am still ok to merge this PR)

ghost commented:
@fclementic2c @alexis-via 
It would be really cool to mark the invoice paid. @pedrobaeza suggested to extend the following method:
https://github.com/odoo/odoo/blob/7682760dcad232829fbb5d4221f66a1ebb9d6c91/addons/account/models/account_invoice.py#L142

in v8 the payments were imported to the bank statement and then reconciled?
https://github.com/odoo/odoo/blob/8.0/addons/account_payment/wizard/account_payment_populate_statement.py 
can we port this wizard to v9 or am I missing something?

Commit:
Merge pull request #3 from sbidoul/patch-1

Avoid confusion between payment mode and payment method

alexis-via commented:
@fclementic2c @BT-kaberer 
I don't understand your point. The supplier invoice is marked as paid when you reach the last step of the payment.order (i.e. the "File Uploaded" state) if you have checked the option "Generate Accounting Entries On File Upload" on the payment mode (it's default value is now True). You can see that in the unitest of the modules account_banking_sepa_credit_transfer (cf https://github.com/akretion/banking/blob/9-sorrento-code-sprint/account_banking_sepa_credit_transfer/tests/test_sct.py#L139) and account_banking_sepa_direct_debit (cf https://github.com/akretion/banking/blob/9-sorrento-code-sprint/account_banking_sepa_direct_debit/tests/test_sdd.py#L140)

So I think there is a mis-understanding somewhere... either I don't understand your point, or you didn't activate the option "Generate Accounting Entries On File Upload" on the payment mode.

ghost commented:
@alexis-via I totally missed that! thank you so much this is awesome ;)

_Cannot generate the account move when the currency of the payment (USD) is not the same as the currency of the company (CHF). This is not supported for the moment._

Is there a workaround for paying in foreign currencies?

sbidoul commented:
@alexis-via thanks for the tests!

One last thing: the uniqueness constraint on account.payment.line seems a bit too strong to me. Was it there before? Should you not add the partner in the key?

sbidoul commented:
@alexis-via I found a bug.

In `wizards/account_payment_line_create.py`, line 116, the test on state != cancel is incorrect:
- because of performance: that `paylines` list will only grow over time
- because it will exclude a move line forever when a partial payment has been done on it

The algorithm in v8 was correct (excluding draft and open), can you use the same?

updated proposed fix in https://github.com/akretion/banking/pull/5

sbidoul committed:
[IMP] prevent deleting a move-line that is linked to a payment order line

sbidoul committed:
[FIX] better filtering of move lines already included in open payment order

sbidoul commented:
@alexis-via regarding this:

> So I think there is a mis-understanding somewhere... either I don't understand your point, or you didn't activate the option "Generate Accounting Entries On File Upload" on the payment mode.

I suggest activating that option by default with the move to the bank account. It's the most intuitive for v9.

Commit:
Merge pull request #5 from acsone/9.0-fix-line-exclusion-sbi

[FIX] better filtering of move lines already included in open payment order

alexis-via commented:
@sbidoul The option "Generate move" is active by default with the move to the bank account (I changed this about 2 weeks ago), cf https://github.com/akretion/banking/blob/9-sorrento-code-sprint/account_payment_order/models/account_payment_mode.py#L56

sbidoul commented:
> @sbidoul The option "Generate move" is active by default with the move to the bank account (I changed this about 2 weeks ago)

Indeed, great!

There is https://github.com/akretion/banking/pull/4 left.

Commit:
Merge pull request #4 from acsone/9.0-restrict-move-line-delete-sbi

[IMP] prevent deleting a move-line that is linked to a payment order line

alexis-via commented:
FYI @BT-kaberer , I plan to finish the work on the multi-currency support in the account move generation on Thursday. So we can merge on Friday :)

sbidoul commented:
@alexis-via argh, I had the finger on the merge button :wink:

alexis-via commented:
@sbidoul well, you can merge this PR, and I'll create a new PR on Thursday. That way, @guewen and other guys who need the account_payment_mode module will be happy.

fclementic2c commented:
yes good idea

sbidoul commented:
Let's go then. Thanks A LOT @alexis-via for this work.
